File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        '''
        time - O(n)
        space - O(n)
        where n is the length of string s
        '''
        if len(s) == 0:
            return 0
        
        l = 0
        count = 1
        char_set = set()
        
        for r in range(len(s)):
            while s[r] in char_set:
                char_set.discard(s[l])
                l += 1
            
            char_set.add(s[r])
            count = max(count, r-l+1)
                
        return count

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        /*
        - longest unique substring
        - substring - subset of a string
        - unique - no duplicate
        - longest - we track state - length
        - how to take substring - 
        - s = "abcabcbb"
        - abca -> bca -> bcab -> cab -> cabc -> abc -> b

        - s = "pwwkew"
        - pww -> w -> wkew -> kew
        - Time - O(NxmaxLength)
        - Space - O(1)
        */

        HashSet<Character> hashSet = new HashSet<>();
        int maxLength = 0;
        int lastIndex = 0;

        for(char sVal:s.toCharArray()) {
            while(hashSet.contains(sVal)) {
                hashSet.remove(s.charAt(lastIndex));
                lastIndex++;
            }
            hashSet.add(sVal);
            maxLength = Math.max(maxLength, hashSet.size());
        }

        return maxLength;
    }
}